### PR TITLE
Refactor perNodeServices options and alllow setting the p2p service type

### PR DESCRIPTION
**Pull request by PierreBesson**

### CODE REVIEW

File: charts/node/templates/service.yaml [hunk 1]
```diff
@@ -56,8 +60,16 @@ metadata:
...
+  {{- if eq $.Values.node.perNodeServices.p2pServiceType "NodePort" }}
   type: NodePort
   externalTrafficPolicy: Local
+  {{- end }}
+  {{- if eq $.Values.node.perNodeServices.p2pServiceType "ClusterIP" }}
+  type: ClusterIP
+  {{- end }}
+  {{- if eq $.Values.node.perNodeServices.p2pServiceType "LoadBalancer" }}
+  type: LoadBalancer
+  {{- end }}
```
Comment (dblane-digicatapult) [inline charts/node/templates/service.yaml:72]: Could use `else if eq` for these and skip the unnecessary `-ends`

File: charts/node/templates/service.yaml [hunk 2]
```diff
@@ -77,8 +89,16 @@ metadata:
...
```
Comment (dblane-digicatapult) [inline charts/node/templates/service.yaml:101]: Could use `else if eq` for these and skip the unnecessary `-ends`

**Review by dblane-digicatapult — CHANGES_REQUESTED** [review]

**Comment by dblane-digicatapult** [comment]
How are you planning to deal with `--public-addr` service address/IP lookup?

I have a rudimentary implementation for my use case (AWS loadbalancer) however this then means advertising `dns4` multiaddr format instead of `ipv4`

The case for loadbalancers is fairly difficult and extremely provider specific and I'm fairly certain there is no "one size fits all" solution.  From memory most of the PaaS providers I tested with advertise the address on either `status.loadbalancer.ingress.ip[]` or `status.loadbalancer.ingress.hostname` So we could try looking up both?

**Comment by PierreBesson** [comment]
> How are you planning to deal with --public-addr service address/IP lookup?

Yes adding the ability to set the loadBalancer IP or hostname would be good. If you want to give it a shot in a following PR it would be very appreciated. We can support any cloud specific way to do this as long as there are users which need this support.

**Comment by dblane-digicatapult** [comment]
> Yes adding the ability to set the loadBalancer IP or hostname would be good. If you want to give it a shot in a following PR it would be very appreciated. We can support any cloud specific way to do this as long as there are users which need this support.

Currently we pass the NodePort `RELAY_CHAIN_P2P_PORT`, `PARA_CHAIN_P2P_PORT` and `EXTERNAL_IP` to the container via a file on the data volume.
Do you want to continue with this approach or should we instead create a configmap which we then do a `kubectl patch` against in the init container? The advantage of using the configmap is that it can be queried by other containers and applications and has potential use for declaring bootnodes and other nodes dynamically.  The disadvantage is that anyone using GitOps tooling to maintain state will have to not overwrite these patches.

Thoughts?

**Comment by PierreBesson** [comment]
>Do you want to continue with this approach or should we instead create a configmap which we then do a kubectl patch against in the init container? The advantage of using the configmap is that it can be queried by other containers and applications and has potential use for declaring bootnodes and other nodes dynamically. The disadvantage is that anyone using GitOps tooling to maintain state will have to not overwrite these patches.

I don't like the approach of using `kubectl patch` directly from the pod. The Kubernetes resource should be the declarative source of truth and should not be modified by init containers. Also GitOps can be pretty much expected. For bootnodes, people should use something like a fixed `nodePort` even though in this case you can only have one replica.

@dblane-digicatapult, it would be better to have this discussion in the issue: https://github.com/paritytech/helm-charts/issues/29

**Review by BulatSaif — APPROVED** [review]

**Pull request merged**